File: 15_Segment-Tree/TAEHO/2042.py
from sys import stdin,setrecursionlimit
setrecursionlimit(int(1e6))

def init(start, end, node):
    if start == end:
        tree[node] = num[start]
        return tree[node]
    mid = (start+end)//2
    tree[node] = init(start,mid,node*2)+init(mid+1,end,node*2+1)
    return tree[node]

def sum(start,end,node,left,right):
    if left > end or right < start:
        return 0
    if left <= start and right >= end:
        return tree[node]
    mid = (start+end)//2

    return sum(start,mid,node*2,left,right) + sum(mid+1,end,node*2+1,left,right)

def update(start,end,node,search,value):
    if search > end or search < start:
        return
    tree[node] += value
    if start == end:
        return
    mid = (start+end)//2
    update(start,mid,node*2,search,value)
    update(mid+1,end,node*2+1,search,value)

n,m,k = map(int,stdin.readline().split())
num = [0]*n
tree = [0] * 4*n

for i in range(n):
    num[i] = int(stdin.readline())

init(0,n-1,1)

for _ in range(m+k):
    a,b,c = map(int,stdin.readline().split())
    if a==1:
        update(0,n-1,1,b-1,c-num[b-1])
        num[b-1] = c
    else:
        print(sum(0,n-1,1,b-1,c-1))

File: 15_Segment-Tree/HONG/2042.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
int N, n, M, K;
ll a, b, c, arr[1000001];

void input() {
    cin.tie(0)->sync_with_stdio(0);
    cin >> N >> M >> K;
    for(int i=1;i<=N;i++) cin >> arr[i];
}

ll init(vector<ll> &tree, int start, int end, int node) {
    if(start == end) return tree[node] = arr[start];
    int mid = (start + end) / 2;
    return tree[node] = init(tree, start, mid, node*2) + init(tree, mid+1, end, node*2+1);
}

ll sum(vector<ll> &tree, int start, int end, int node, int left, int right) {
    if(left > end || right < start) return 0;
    if(left <= start && end <= right) return tree[node];
    int mid = (start + end) / 2;
    return sum(tree, start, mid, node*2, left, right) + sum(tree, mid+1, end, node*2+1, left, right);
}

void update(vector<ll> &tree, int start, int end, int node, int target, ll diff) {
    if(target > end || target < start) return;
    tree[node] += diff;
    
    int mid = (start + end) / 2;
    if(start != end) {
        update(tree, start, mid, node*2, target, diff);
        update(tree, mid+1, end, node*2+1, target, diff);
    }
}

int main() {
    input();

    vector<ll> tree(1 << (int)ceil(log2(N) + 1));
    init(tree, 1, N, 1);

    while(cin >> a >> b >> c)
        if(a==2) cout << sum(tree, 1, N, 1, b, c) << '\n';
        else {
            update(tree, 1, N, 1, b, c-arr[b]);
            arr[b] = c;
        }
}
